### ISSUE TITLE
Quantum boogies - Quantum neural distinguisher for differential cryptanalysis

### ISSUE THREAD

**Issue by khj1594012** [qiskit-community/quantum-hackathon-korea-22#2]
# Abstract

We implemented a quantum neural distinguisher for S-DES using a quantum-classical hybrid neural network and compared it with the classical neural distinguisher. In all cases, the quantum-classical hybrid neural network performed better than the classical neural network, and has the advantage of achieving better performance with fewer epochs and data. Also, when the classical neural network fails in the problem of classifying random data and ciphertext data satisfying a specific difference, the hybrid neural network succeeds. In our knowledge, this is the first work on a neural distinguisher based on quantum neural networks.

# Members

 - @khj1594012
 - @sejinlim1107
 - @kangsquad
 - @wwoong
 - Qiskit Coach: @

# Introduction 
Differential cryptanalysis is one of cryptanalysis for block ciphers. It is a technique to infer the key by analyzing the output value according to the change of the input value when the substitution of the cipher is designed to be weak. It works more effectively compared to the brute force. Differential cryptanalysis attempts to attack through a specific plaintext pair called a neural distinguisher that satisfies the difference. Studies on differential cryptanalysis based on artificial neural networks have been conducted[Gohr, Baksi], because of time and resources wasted to find a plaintext pair that satisfies the difference one by one. In this hackathon project, we implemented a quantum neural distinguisher using a quantum-classical hybrid neural network, and applied it to S-DES encryption. In our knowledge, this is the first work on a neural distinguisher based on quantum neural networks. Little research has been done on cryptanalysis using quantum neural networks, and [Kim] is the result of performing known-plaintext attacks.

# Related work 
## Classical Neural Network 
Artificial neural networks are learning algorithms inspired by neural networks in biology. As shown in Figure 1, a neural network is constructed in the form of stacked layers of multiple nodes. As shown on the right of the Figure 1, neurons (i.e. nodes) in each layer perform a weighted sum operation using the node values (x) and weights (w) of the previous layer connected to them, and add a bias. Then, it is input to the non-linear activation function, and computed as a single value. In this way, the loss value is obtained after passing through all the layers. Then, the weights inside the neural network are updated to minimize the loss through the backpropagation process. By repeating this process, a neural network that guarantees generalization performance for untrained data is constructed. When the trained model is used for actual inference, the inference proceeds by inputting data with the weights of the  fixed neural network. Through this, it is possible to learn, classify, and predict by extracting features of input data (e.g. image, time-series, language, and graph).

<div align=center><img width="600" alt="image" src="https://user-images.githubusercontent.com/57383955/176279813-82f8c504-c71a-4e78-8260-65b922696c3a.png"><br>
Figure 1. Architecture of artificial neural networks.</div>

## Quantum Neural Network 
A quantum neural network is an artificial intelligence that utilizes quantum mechanics phenomenon (entanglement and superposition). A quantum neural network consists of qubits and quantum gates on a quantum computer. Therefore, it learns quantum state data (parameterized quantum circuit) by encoding the classical data into quantum data. The parameter of the circuit is the rotation angle of the rotation gate, and the input data is used as the rotation angle. After encoding, the constructed quantum circuit is executed. At this time, each qubit state is changed as it passes through the quantum gate. Finally, when a qubit is measured, the state of the qubit is determined to be a classical value.  The loss is calculated based on that value and the rotation angles of the quantum gate are changed. The parameterized quantum circuit is trained by re-executing the quantum circuit to which the changed parameter is applied and repeating this process. Quantum neural networks require fewer parameters and fewer training data than classical neural networks. It also has the advantage of being able to perform better than classical neural networks. 

## Quantum Classical Hybrid Neural Network 
A classical-quantum hybrid neural network is a neural network that uses a combination of classical and quantum neural networks. In other words, a parameterized quantum circuit is used as one of the layers that make up a classical neural network. Therefore, elements such as loss functions, optimization functions, metrics, and epochs are used the same as in classical neural networks. The training process consists of calculating the loss as a forward propagation process like a classical neural network and adjusting parameters through a backpropagation process based on the loss value. As such, the whole process is the same as for classical neural networks, but with quantum circuits as one layer. In quantum circuit execution, the state of qubits is measured after sequentially passing all gates constituting the circuit in the same way as in a quantum neural network. Finally, the expected value of the quantum circuit is obtained and used as the final output of the quantum layer. After calculating the loss based on the corresponding value, the existing parameter (the rotation angle of the qubit) is changed through a parameter shift method, etc. Through this process, the hybrid neural network is trained. Currently, there are many studies using a hybrid neural network, and it has the advantage of being more stable and performing better than a neural network using only quantum. 

## Differential Cryptanalysis 
Differential cryptanalysis is a representative cryptanalysis method of block ciphers. The change of the output value according to the change of the input value is predicted and analyzed with a certain probability. A linear layer is necessarily predictable, but a non-linear layer is probabilistically predictable. The S-box used in cryptographic algorithms is a non-linear process. When finding a key through differential cryptanalysis, if the key is correct, the input/output difference will be satisfied with a high probability. This means that there are cases where a certain difference is satisfied. Conversely, if the key is wrong, the relationship between the plaintext and the ciphertext will disappear, resulting in a random probability. The more chosen plaintext we have, the better we can distinguish the correct key. Also, the more we know the plaintext pairs required for differential cryptanalysis, the higher the probability of recovering the key. The differential cryptanalysis process consists of three steps. In step 1, we find the differences. In step 2, we find a pair of plaintexts that satisfy the difference found in step 1. Finally, differential cryptanalysis is performed by brute force on the round key for r rounds. In the brute force process for the round key, after obtaining the ciphertext for round r+1, the ciphertext is decrypted for one round through brute force. If the decoded result satisfies the output difference of r rounds, it is a correct round key. Conversely, if the output difference is not satisfied, the key is incorrect. After finding the round key in this way, the entire key can be found. Additionally, the step 2 process is performed with a neural distinguisher implemented through an artificial neural network.

## Simplified DES (S-DES) 
S-DES is a simplified cipher of DES, and it is a cipher that uses the same encryption method while simply reducing the block size, key size, and the number of rounds. Table 1 shows the details of S-DES.

<img width="849" alt="image" src="https://user-images.githubusercontent.com/57383955/176283296-564738c3-0e9f-4083-bb10-9f294fb5fdb8.png">

## Neural Distinguisher 
The neural distinguisher is used for the process of finding a plaintext pair that satisfies the difference required for differential cryptanalysis (Step 2). It performs the task of discriminating between random data and ciphertext pairs that satisfy the difference. That is, it is a binary classification of random data and ciphertext data satisfying the difference. If there is only one input difference, it is binary classification and should achieve an accuracy of 0.5 or better. When classified as random data, it is not used as a distinguisher because it is not a plaintext pair that satisfies the difference, and in the opposite case, it is used as a distinguisher for differential cryptanalysis.

# Our work 
## Quantum Neural Distinguisher with Hybrid Neural Network 
In this hackathon project, we implemented a quantum neural distinguisher for S-DES using a quantum-classical hybrid neural network and compared it with the classical neural distinguisher. The quantum neural distinguisher is an implementation of the classical neural distinguisher as a quantum neural network. To the best of our knowledge, quantum neural distinguishers have not yet been studied. Figure 2 shows the system diagram. First, plaintext pairs that satisfies the difference and a random plaintext pair are generated. The input differential used is 0x96 [S-DES differential]]. Corresponding plaintext pairs are encrypted using S-DES, ciphertext pairs are labeled 1, and random data are labeled 0. Figure 3 shows the details of dataset. These data are fed into the hybrid neural network. A hybrid neural network is a neural network that combines a classical neural network and a quantum neural network. That is, the quantum circuit is used as one layer of the neural network. Input data is input to a classical neural network, features are extracted and then input to a quantum circuit. The quantum circuit uses the input value as the rotation angle. After operating the quantum circuit, it is determined as a single value through measurement. And, the parameters are adjusted using the calculated loss values. Through this process, random data and data satisfying a difference can be classified. Data satisfying the difference are used for differential cryptanalysis, and random data is discarded.

<div align=center><img width="600" alt="image" src="https://user-images.githubusercontent.com/57383955/176283969-62332b2c-69a7-45be-b1e2-055a3e4209cd.png"><br>
Figure 2. System diagram.</div><br>

Figure 4 shows the overall structure of the hybrid neural network. It is a form in which a quantum circuit is added to a classical input layer and a classical hidden layer. After running the quantum circuit, calculate the count for each state and find the probability. Through this process, the expectation for 1 is obtained, so ‘torch.cat((x, 1 - x), -1)’ is performed to apply the loss function. This value becomes the final output of the hybrid neural network. Figure 5 shows the quantum circuit used as the quantum layer. A single qubit and 'ry' rotation gate were used. The number of neurons in the hidden layer of the classical neural network is 1, and the layer outputs a single value. The corresponding value is used as the rotation angle of the rotation gate. Thereafter, after changing a parameter through a parameter shift method, the quantum circuit is operated again. In this way, it learns while updating the parameters. When operating a hybrid neural network, hyperparameters can be changed (e.g. epochs, shots, learning rate, etc.).<br><br>

<div align=center><img width="550" alt="image" src="https://user-images.githubusercontent.com/57383955/176284316-b1bd9251-32df-449d-9db4-643e8b542431.png"><br>
Figure 3. Dataset</div><br>

<div align=center><img width="600" alt="image" src="https://user-images.githubusercontent.com/57383955/176284406-6cf94ca5-2095-4845-adc4-08fba1f6384b.png"><br>
Figure 4. Architecture of hybrid neural network.</div><br>

<div align=center><img width="600" alt="image" src="https://user-images.githubusercontent.com/57383955/176284604-b624f418-c429-4058-8256-a09042e4b659.png"><br>
<img width="600" alt="image" src="https://user-images.githubusercontent.com/57383955/176284614-efa39738-d449-4494-b70e-a40b9d9d4738.png"><br>
Figure 5. Architecture of quantum circuit (quantum layer).
</div>

# Evaluation 
Our experimental environment is as follows. 
- IBM quantum lab and Google Colab 
- Qiskit + pytorch 
- Aer simulator

<div align=center><img width="600" alt="image" src="https://user-images.githubusercontent.com/57383955/176284993-fd99169c-fe08-4bca-b23d-6bedea137e7e.png"><br>
Table 2.  Comparison of classical neural distinguisher and quantum neural distinguisher for S-DES</div><br>
We compared the neural distinguisher implemented by the classical neural network and the hybrid neural network. In both neural networks, the classic part has the same structure, and in the hybrid neural network, a quantum circuit with 1-qubit is added at the end.

## The number of data : 500 
First, classical neural networks failed to classify. However, in the case of quantum hybrid neural networks, accuracy exceeding 50% was achieved when shots were 100. In addition, since 52% accuracy was achieved even with fewer epochs, we succeeded in discriminating ciphertext pairs that satisfy the difference. That is, it can be successfully used in the process of finding a plaintext pair that satisfies the input difference corresponding to step 2 of the differential cryptanalysis.

## The number of data : 1000 
Next, the classical neural network achieved an accuracy of 53.5% on the test data. In the hybrid neural network, we achieved an accuracy of 57.3% for 100 shots and 50 epochs, and 57.5% accuracy for 100 epochs. Also, when the epoch was reduced to 30, an accuracy of 56.9% was obtained, which is also higher than that of the classical neural network. Finally, when the shots were increased to 1024, the highest accuracy of 58.8% was achieved. Through this, we can see that as the value of shots increases (appropriately), the accuracy improves.

## Quantum advantages for differential cryptanalysis 
The quantum advantages of using a hybrid neural network are as follows. In all cases, the quantum-classical hybrid neural network achieved higher accuracy than the classical neural network. In other words, better performance can be achieved with fewer epochs and fewer training data. In addition, in the classical neural network, there were cases where the classification failed because the accuracy did not exceed 50%, but the hybrid neural network succeeded in classifying the same case.

# Conclusion and Future Works 
We implemented a quantum neural distinguisher using a hybrid neural network and compared it with a classical neural network. The case of using a hybrid neural network has a quantum advantage in that higher accuracy can be obtained using less training data and fewer epochs than a classical neural network. In addition, tasks that were not successful with classical neural networks were succeeded with hybrid neural networks. Therefore, this quantum neural distinguisher can be successfully used for differential cryptanalysis for S-DES. In future work, we will try to perform cryptanalysis for other cipher (e.g. S-AES, DES, Speck), and improve our accuracy. Also, we will apply QSVM, which consists of only quantum circuits, to a quantum distinguisher.

# References 
[Gohr] Gohr, Aron. "Improving attacks on round-reduced speck32/64 using deep learning." Annual International Cryptology Conference. Springer, Cham, 2019. 
[Baksi] Baksi, Anubhab. "Machine learning-assisted differential distinguishers for lightweight ciphers." Classical and Physical Security of Symmetric Key Cryptographic Algorithms. Springer, Singapore, 2022. 141-162. 
[Kim] Kim, Hyun-Ji, et al. "Cryptanalysis of Caesar using Quantum Support Vector Machine." 2021 IEEE International Conference on Consumer Electronics-Asia (ICCE-Asia). IEEE, 2021. 
[S-DES differential]Raj, Keshav, et al. "Differential cryptanalysis on S-DES." International  Journal of Management & Information Technology 1.2 (2012): 42-45.

# GitHub repo

https://github.com/khj1594012/Quantum-Neural-Distinguisher-for-S-DES

**Comment by ZiJUEWang** [comment]
请问有这篇文章的代码吗，想学习参考一下，谢谢
